File: ta_lib/strategies/pulse/src/nvol.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;
use volume::nvol;

const NVOL_LINE: f32 = 100.0;

pub struct NvolPulse {
    smooth: Smooth,
    period: usize,
}

impl NvolPulse {
    pub fn new(smooth: Smooth, period: f32) -> Self {
        Self {
            smooth,
            period: period as usize,
        }
    }
}

impl Pulse for NvolPulse {
    fn lookback(&self) -> usize {
        self.period
    }

    fn assess(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let nvol = nvol(data.volume(), self.smooth, self.period);

        (nvol.sgt(&NVOL_LINE), nvol.sgt(&NVOL_LINE))
    }
}

File: strategy/generator/pulse/nvol.py
from dataclasses import dataclass

from core.models.parameter import CategoricalParameter, Parameter, StaticParameter
from core.models.smooth import Smooth

from .base import Pulse, PulseType


@dataclass(frozen=True)
class NvolPulse(Pulse):
    type: PulseType = PulseType.Nvol
    smooth: Parameter = CategoricalParameter(Smooth)
    period: Parameter = StaticParameter(14.0)
